Add script to flatten git history for llvm monorepo submission

peterwaller-arm opened the pull request:
## [Click here for the latest description of intent posted to flang-dev](http://lists.llvm.org/pipermail/flang-dev/2019-December/000122.html)

Please note this pull request is not intended to be merged, it is here for review only.

Please see `flatten.cpp`, which contains significant self-documentation. `flatten.sh` was an earlier shell script with a similar intent, but is now obsolete, and just left here for reference.

In summary, this program takes as input the upstream llvm-project monoreo and the f18 project. As output, it generates from the f18 history a new series of commits following on from the HEAD of the llvm-project repository. Additionally, the commits are rewritten so that it appears as though all of the project history happened inside the `flang/` directory.

The key reason to do this is to eliminate merge commits. Secondarily, we improve the commit messages so that references to pull requests and issues point back into this repository correctly and will continue to work into the future, rather than referencing arbitrary unrelated issues llvm-project.

To achieve the flattening we follow the first parent history of f18. When a merge commit is encountered, the second-parent is rebased onto the first-parent. This is of course "dangerous", because the commits on the second-parent lineage will not have the same checkout as the original commits. To mitigate this, we make certain that by the end of the rebase, the checkouts are the same as the original master.

There are a few cases where they are not, and in those cases the second-parent history is squashed into a single commit.

However, it is guaranteed that the checkouts at the f18 merge commits will be equivalent, within the flang directory.

tskeith commented on flatten.sh:
```diff
@@ -0,0 +1,118 @@
+# * Individual commits within PRs are lost. This is somewhat fundamental, and
+#   could be worked around with some effort; it would require rebasing a number
+#   of patches, which is risky and time consuming, so instead the working tree
+#   is taken as-is from the merge commit.
```
I think this is a serious short-coming. Squashing together all of the commits in a pull request can lose a lot of information. See the new commit 5e2cd5eb066062b7bd6d7874817d5213187bccb2 for an example. It combines several commits, some of are refactorings and some are function changes.

peterwaller-arm replied:
I agree that isn't great.

The only alternative I'm aware of is to do effectively an interactive rebase, which would also flatten history, discarding only the (otherwise empty) merge commits. As far as I can tell, that would require some effort to deal with various bits of the history where commits can't be replayed without the merge commit. I don't see why this couldn't be done in principle, but it would require additional engineering effort.

RichBarton-Arm replied:
@peterwaller-arm mentioned to me this evening that he might have a mitigation to this in that the new squashed commits could link back to the original F18 git shas that went into them. So the F18 git history in LLVM monorepo is straightline (and cheaply done) but has some commits that combine things together but with the original commits, that can be clicked through to in github, to help separate them.

peterwaller-arm replied:
The next update does a better job. Now there are only 6 merge commits which get squashed. As described elsewhere, fixing those would require manual work. Please take another look.

peterwaller-arm commented:
I've posted an update here: http://lists.llvm.org/pipermail/flang-dev/2019-December/000105.html

Key changes:
* New way of preserving commits, such that we preserve many more commits.
* Added some trailer headers and `Reviewed-on` links.

The script got quite a bit more complicated, but I think we can have a base level of confidence that the history is not significantly messed up because the trees of the mainline commits are the same.

There are a couple of known deficiencies:

* There are 66 commits from 6 merges which have been sqaushed into 6 commits. As far as I currently know, fixing this would require a manual rebase of those 66 commits.
* Original-Commit is not consistently set, it's missing for commits which have been rebased. This is a smallish bug that could be fixed.
* There are some "merge commits" (with 1 parent) in the rewritten history which complicate things. The history is still linear, it just has these extra commits. I haven't yet had time to think about what to do with them but I believe there is likely to be a simple solution.

tskeith commented on flatten.sh:
```diff
@@ -0,0 +1,277 @@
+            git show --no-patch --format=%B "$SHA"
+            # echo Original-Commit: flang-compiler/flang@${SHA}
+            make_reviewed_on ${merge_commit}
+        } > message.txt
```
I have one small suggestion here: It's good to have the link back to the original pull request, but I think it should be easier to tell which PR a commit is from, for example, in the output of `git log --oneline`. We've lost the information that the commits were grouped together in a pull request.

My suggestion is to prefix the title with something like "PR123: " when the commit comes from pull request 123.

peterwaller-arm replied:
In the latest flatten.cpp now pushed, I wrote it as `[flang-compiler/f18#NNN]` so that the cross reference will not reference the LLVM repository, but this repository.

peterwaller-arm commented:
I have pushed an update including the C rewrite.

The full description is in the [mailing list post](http://lists.llvm.org/pipermail/flang-dev/2019-December/000122.html).

Other key links:


* Tree, merged with LLVM: 
https://github.com/peterwaller-arm/f18/tree/rewritten-history-v2-llvm-project-merge

* Rewritten history: 
https://github.com/peterwaller-arm/f18/commits/rewritten-history-v2-llvm-project-merge

* Rewritten history without llvm merge: 
https://github.com/peterwaller-arm/f18/commits/rewritten-history-v2

peterwaller-arm commented:
@tskeith: Aaron Smith requested on the llvm-dev mailing list that the commit subjects don't include `[flang-compiler/f18#123]` but just say `[flang]`, with the rationale that nothing is lost -- and his comments come after we now retain the merge commits. Do you think we can change the prefix to `[flang]` now? Individual commits won't show the PR number in the title, but it should be clear from the position of the empty merge commits, and the commit message body does have the "Reviewed-on" trailer.

tskeith commented:
> @tskeith: Aaron Smith requested on the llvm-dev mailing list that the commit subjects don't include `[flang-compiler/f18#123]` but just say `[flang]`, with the rationale that nothing is lost -- and his comments come after we now retain the merge commits. Do you think we can change the prefix to `[flang]` now? Individual commits won't show the PR number in the title, but it should be clear from the position of the empty merge commits, and the commit message body does have the "Reviewed-on" trailer.

So the history would look like one or more commits that went into PR 123 followed by an empty commit titled "Merge pull request flang-compiler#123 from ...", right? That seems fine to me.

But looking at your rewritten history, there seem to be a few that don't look like that. For example, there is a commit for PR #867 but no merge commit. Why is that?

peterwaller-arm commented:
> But looking at your rewritten history, there seem to be a few that don't look like that. For example, there is a commit for PR #867 but no merge commit. Why is that?

```
$ git log --oneline --graph --decorate b3bb698086a..3d05d39e363
*   3d05d39e363 Merge pull request #868 from schweitzpgi/remove-flatten
|\  
| * d7554c76050 remove some old files
* | 61bb9db531a Flang relicense: LICENSE file and one change (#867)
* | ae4868e572c Updated call10.f90 with more test for #794-Check for PURE procedures containing image control statements (#857)
* |   53545d7436c Merge pull request #865 from flang-compiler/jpr-fix-cplx-power
|\ \  
| * | 9a84bcb9ac3 Fix INTGER**COMPLEX regression introduced by PR 858 Fix bad copy past that had introduced an unitialized value use.
|  /  
* | 4771f8cb789 Fix file path in cmake (#844)
|/  
* d1b6e3cd2c4 Merge pull request #864 from flang-compiler/pmk-fix
* e58c4e53c2f Revert recent addition of ActualArgument::PassedObject
```

The above hitstory shows the commit which "merged" 61bb9db531a is actually a cherry-pick/squash commit. So where github shows "merged" in the user interface, I believe those PRs get a commit message on master which looks like `<Pull request title> (#<PR number>)`, and not actually a merge commit. There are a fair number of those in that form.

That is actually a case where I haven't generated "Reviewed-on" headers, because they aren't distinguishable from commits where someone writes `(#<issue number>)` from the history alone, I would also need to obtain a list of numbers which are pull requests to achieve that.

peterwaller-arm commented:
> I would also need to obtain a list of numbers which are pull requests to achieve that.

For now I don't propose to fix up these edge cases, unless it seems particularly important. I have the impression they are relatively rare, but they do represent a gap of some unknown size in the `Reviewed-on` trailer headers.

peterwaller-arm commented:
I got a few minutes at a computer. There are 32 such squash commits. I was able to verify using the github API that all commits with subjects ending in "(#[0-9]+)" are indeed pull requests.

One-liner to grab PR numbers: `curl https://api.github.com/repos/flang-compiler/f18/pulls?state=all'&'page='[1-23]' | grep -v '_curl_' | jq -c -s '.[][].number'`

I'm assuming no such commits will land between now and when we do a rewrite which name an issue number in that position rather than a PR, but the worst that happens is that we generate incorrect Reviewed-on lines for them.

A few straightforward changes above:

* These commits now have a Reviewed-on reference.
* Subjects no longer begin `[flang-compiler/f18#nnn]`
* I switched the sense of the tree-same-pre-rewrite header. Now only commits which don't result in the same checkout are marked. I thought this was obviously the better way around in hindsight, since it means there is a hint there on the commits which need it that they aren't equal. It also makes them easier to search for by grepping the commit log in the pager. We have't discussed this header very much. I think it is nice to leave a hint that these commits aren't quite what they used to be before the rebase. Hopefully if it ever comes up, it is recognized as such and searched for, which leads back to this PR. We could just drop it if it nobody else thinks this is useful.

peterwaller-arm commented:
And here's the latest history:

https://github.com/peterwaller-arm/f18/commits/rewritten-history-v3

peterwaller-arm commented:
Two proposed changes:

* I note that the MLIR commit was merged in as a second root, so llvm/llvm-project@aed0d21a62d does not have any parents. We could do the same. See [http://lists.llvm.org/pipermail/llvm-dev/2020-January/137960.html](this post for further detail).

* There was a suggestion to prefix flang commits with `[flang]`, which seems reasonable to me. I removed `[flang-compiler/f18#nnn]` but forgot to re-add that.

peterwaller-arm commented:
I've added a few commits above to tweak minor things that came up. Now we prepend all the flang commits with `[Flang]`, as is considered polite in LLVM.

The program now generates the merge commit itself. So the program now directly produces the commit we intend to push into LLVM.

In contrast to previously where we based the flang history on top of the latest llvm history, the latest program now generates a zero-parent first commit. The effect of this is that if I pick a random flang commit [it only has the flang folder in it](https://github.com/peterwaller-arm/f18/tree/9bafcc671d2c08ce4ce7f5769e82b7ecd9722c55):

![Screenshot from 2020-01-09 14-04-36](https://user-images.githubusercontent.com/52030119/72074007-01f7b080-32e9-11ea-8c8c-e7ba5a13ae65.png)

We don't have to do it this way, but it is how MLIR did it, and it makes some sense to me because up to that point the trees were independent. I raised this on the mailing list [but no-one commented either way](http://lists.llvm.org/pipermail/llvm-dev/2020-January/137960.html). I think it's best to follow MLIR's lead.

#### Here is the new flang history lineage (second-parent of the merge): https://github.com/peterwaller-arm/f18/commits/rewritten-history-v4-llvm-project-merge%5E2

![Screenshot from 2020-01-09 14-05-31](https://user-images.githubusercontent.com/52030119/72074075-25baf680-32e9-11ea-8c7d-5701587e7592.png)

#### Here is the llvm lineage after the merge: https://github.com/peterwaller-arm/f18/commits/rewritten-history-v4-llvm-project-merge -- note that git's display of commits appears to sort them in date order, rather than topologically, so the flang commits appear interleaved with the llvm commits:

![Screenshot from 2020-01-09 14-06-08](https://user-images.githubusercontent.com/52030119/72074110-35d2d600-32e9-11ea-9873-9315d814d9d8.png)

[...]

![Screenshot from 2020-01-09 14-06-37](https://user-images.githubusercontent.com/52030119/72074160-497e3c80-32e9-11ea-845b-ba540ddb04a7.png)

peterwaller-arm commented on flatten.cpp:
```diff
@@ -0,0 +1,916 @@
+            "\n"
+            "Updates: flang-compiler/f18#876 (submission into llvm-project)\n"
+            "Mailing-list: http://lists.llvm.org/pipermail/llvm-dev/2020-January/137989.html ([llvm-dev] Flang landing in the monorepo - next Monday!)\n"
+            "Mailing-list: http://lists.llvm.org/pipermail/llvm-dev/2019-December/137661.html ([llvm-dev] Flang landing in the monorepo)\n";
```
@tskeith This is the commit message of the merge. I chose to include some extra references here so that people unfamiliar with what's going on have something to go by when they see the merge in their histories. Please review this, which is new.

tskeith replied:
> @tskeith This is [the commit message](https://github.com/peterwaller-arm/f18/commit/527fdd0649249bd076681cf0c2230eb096832f7f) of the merge. I chose to include some extra references here so that people unfamiliar with what's going on have something to go by when they see the merge in their histories. Please review this, which is new.

That looks okay to me.

RichBarton-Arm commented:
@sscalpone Do you think it makes sense to merge this PR into master so we have the script available in source code when the upstreaming to LLVM time comes?

peterwaller-arm commented:
In the above force push:

* Rebased against current master.
* Removed obsolete flatten.sh.
* Squashed the history to a single commit.

I re-ran the script against the current llvm-project and f18 HEADs and everything continued continues to pass the sniff test.

RichBarton-Arm commented:
@sscalpone - this is ready to merge.
